include/drivers/rtc.h

```diff
@@ -0,0 +1,67 @@
+/*
+ * Copyright (c) 2020 Paratronic
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#ifndef ZEPHYR_INCLUDE_DRIVERS_RTC_H_
+#define ZEPHYR_INCLUDE_DRIVERS_RTC_H_
+
+/**
+ * @file
+ * @brief Public RTC driver APIs
+ */
+
+#include <zephyr/types.h>
+#include <time.h>
+#include <device.h>
+
+typedef int (*rtc_api_get_time_t)(struct device *dev,
+				  struct timespec *tp);
+typedef int (*rtc_api_set_time_t)(struct device *dev,
+				  const struct timespec *tp);
+
+__subsystem struct rtc_driver_api {
+	rtc_api_get_time_t get_time;
+	rtc_api_set_time_t set_time;
+};
+
+/**
+ * @brief Function to get RTC time
+ *
+ * @param[in] dev RTC device
+ * @param[out] tp pointer to where to store the current rtc value
+ *
+ * @return 0 on success, negative on error
+ */
+__syscall int rtc_get_time(struct device *dev, struct timespec *tp);
+
+static inline int z_impl_rtc_get_time(struct device *dev, struct timespec *tp)
+{
+	const struct rtc_driver_api *api =
+		(const struct rtc_driver_api *)dev->driver_api;
+
+	return api->get_time(dev, tp);
+}
+
+/**
+ * @brief Function to set RTC time
+ *
+ * @param[in] dev RTC device
+ * @param[in] tp pointer to new rtc value
+ *
+ * @return 0 on success, negative on error
```

Comment:
This is probably the minimal necessary.  For precision timing you want a way to correlate the preserved time with the local system clock in ticks.  https://github.com/zephyrproject-rtos/zephyr/pull/17631 has some API that does that.

Reply:
This synchronization is interesting. In my opinion, it must be done regularly in the driver (according to the drift of system clock) to permit the [get / set] _time functions to have more precision. I don't think that exposed some synchronization API is necessary.

Reply:
I don't think adjustment belongs in the RTC driver.  The local RTC itself may be corrected based on an external time source like PTP, NTP, GPS, or mesh time service, and many will have device-specific mechanisms to tune the rate.

The ability to align the system tick clock with each time source independently (for both sets and gets) allows you to transfer times from one source to another with minimal error and without burdening each time-related component with extra code trying to compensate for anticipated error.

The tick drift may not be constant (temperature effects) so I expect it to be maintained separately with its own infrastructure, retrieving synchronized time values from different sources and adjusting the tick-to-time offset parameters dynamically.  We're not at that level of maturity in Zephyr yet.

Reply:
Thank you for this exhaustive response. It is true that in my current case, I wish to have only one precise time reference to be able to timestamp my data. My approach to the subject is therefore surely a little simplistic. I don't have enough experience in clocks managing and synchronization mechanisms to know which structure would be suitable for this.